### PR TITLE
Add some guards for base community logic

### PR DESCRIPTION
**Pull request by benhalpern**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
Adds the concept of `base_community?` to check if the site is `dev.to`. This can be used to guard against including features in additional communities that we only have implemented in DEV and don't yet have a relevant `SiteConfig`.

These should generally be seen as areas for TODO in generalization, but in a couple places there may always be a differentiation. The footer message is one place.

This is not an exhaustive implementation, but creates the functionality and drops it in a few key places.

I left another comment below going into more details about intended purpose here.

### CODE REVIEW

File: app/views/layouts/_footer.html.erb [hunk 1]
```diff
@@ -3,7 +3,14 @@
     <div class="inner-footer-container">
       <a href="/">Home</a> <a href="/about">About</a> <a href="/privacy">Privacy Policy</a>
       <a href="/terms">Terms of Use</a> <a href="/contact">Contact</a> <a href="/code-of-conduct">Code of Conduct</a><br/>
-      <%= ApplicationConfig["COMMUNITY_NAME"] %> Community copyright 2016 - <%= Time.current.year %>&nbsp; 🔥
+      <%= ApplicationConfig["COMMUNITY_NAME"] %> Community copyright <%= "2016 -" if base_community? %> <%= Time.current.year %>&nbsp; 🔥
```
Comment (JibranKalia) [inline app/views/layouts/_footer.html.erb]: Does it make sense to add a todo somewhere that `ApplicationConfig["COPYRIGHT_START_YEAR"]` has to be defined in the future?

**Comment by citizen428** [comment]
> These should generally be seen as areas for TODO in generalization, but in a couple places there may always be a differentiation. The footer message is one place.

I don't think that necessarily needs to be true. @rhymes and I talked in the past about adding I18n support to the app, which even makes sense in the context of only supporting a single language (e.g. copy can change without breaking tests if specs assert on `I18n.t(...)`. A lot of these messages could be summarized under a common key in the translation files and therefore easily be customized by each community. something like

```yml
en:
  community:
   footer_message: Built transparently...:
```

**Comment by rhymes** [comment]
Replying to @citizen428 here:

> I don't think that necessarily needs to be true. @rhymes and I talked in the past about adding I18n support to the app

I'm ccing @jacobherrington as I was referencing his issue here: https://github.com/thepracticaldev/dev.to/issues/3929

>, which even makes sense in the context of only supporting a single language (e.g. copy can change without breaking tests if specs assert on `I18n.t(...)`). A lot of these messages could be grouped under a common key in the translation files and therefore easily be customized by each community. 

Can they customize them without having to fork the code? Because if they edit the translation file they'd have to commit it somewhere, correct?

> I strongly believe that with the generalization effort progressing, the default should be the generalized version, without any references to a base community.

Agreed. The more we can ourselves be just _an installation_ of this code, without being a special case for it, the better

**Comment by benhalpern** [comment]
Soooo my thoughts for this are to help us deal with certain parts of the app which aren't really "of quality" to be part of the generalization effort, such as on-platform video right now since it's feeble as an end-user feature but we do use it anyway. I just wouldn't want to infect new communities with half-baked features if we can just hide them for now.

And then there's the _currently hardcoded_ stuff that is kind of optional in nature (like those DEV logos). I think the best case is that we have full-on replaceability for key platform images but in the meantime we patch it like this wherever we find it.

There are only a few places where it might "live on forever" in any way, but I think that's where these better solutions come into play. E.g. this is kind of designed as a patch tool as we go live and encounter issues to be systematically replaced by better approaches.

**Comment by citizen428** [comment]
@rhymes 

> Can they customize them without having to fork the code? Because if they edit the translation file they'd have to commit it somewhere, correct?

Good point. There are tools that let you pull in translations from web services like WebTranslateIt, so theoretically the file could be pulled in from elsewhere during deploy. We could probably roll a similar system ourselves. Or just make the relevant strings part of `SiteConfig` or a similar system. Anyway, we still have time to think about this.

**Comment by citizen428** [comment]
> Is "development & release of new features on DEV codebase, but only release them for all communities if they're solid" a process we'll be sticking with in the long run?

In the interest of generalization, I would say major new features should almost always go behind a feature flag/configuration options, so other communities can decide if they want to use it all and also how soon they want to use it (first release, once the future proved itself on other communities etc).

**Comment by rhymes** [comment]
Feature flags is a topic that we definitely need to figure out sooner or later. I don't think `SiteConfig` is that (that's just about global configuration, even the "twitter or github" auth enabling might eventually become a feature flag) and I believe we should have something more "generalized" than a binary "edge feature, yes or no" (or base community yes or no). We did have the [sail gem](https://github.com/vinistock/sail) in at some point but IIRC we didn't make much use of it, maybe we can reconsider something like it.

I'm not advocating for a full fledge "feature flag" framework but we can definitely cook something up or find something to suit our needs. I still think the binary boolean is limiting as feature flags should definitely be something that we can employ even on DEV. I see that as true generalization.

What if we consider ripping the bandaid off on feature flags in the next cycle and we consider Ben's changes from this PR part of that too? Don't know how granular we want to go but I think we can figure that out en route.

**Comment by citizen428** [comment]
@rhymes You raise some good points. Suggestion: we take the generalization work form cycle 2 as input for what we expect of a feature flag framework, evaluate the available ones, sketch out what our implementation could look and then tackle the issue itself in cycle 3. WDYT?

**Comment by lightalloy** [comment]
I have just looked at the pr and discussion today. I agree that eventually, we should have some kind of feature enabling/disabling. I18n would also be useful, but it's another topic.
I suppose that we'll be able to figure out a more clear approach by working on the generalization in cycle 2 (so, agree with @citizen428 on the point from the last comment)

**Comment by mstruve** [comment]
Lots of discussion happening here which is awesome but now it looks like most of these changes are now out of date. I am going to close this for now bc of all the conflicts. Feel free to update and reopen Ben!!!

**Pull request closed without merging**